src/cow-react/modules/trade/hooks/setupTradeState/useSetupTradeState.ts

```diff
@@ -1,195 +1,168 @@
+      }
 
-    updateState?.(newState)
+      return
+    }
 
-    tradeNavigate(newState.chainId, {
-      inputCurrencyId: newState.inputCurrencyId || null,
-      outputCurrencyId: newState.outputCurrencyId || null,
-    })
-  }, [tradeNavigate, newChainId, chainIdFromProviderWasChanged, state, updateState, tradeStateFromUrl])
+    updateState?.(tradeStateFromUrl)
+    console.debug('[TRADE STATE]', 'Applying a new state from URL', tradeStateFromUrl)
 
-  /**
-   * STEP 1
-   * Unlock network switching in the provider when chainId was changed in the URL
-   */
-  useLayoutEffect(() => {
-    if (chainIdFromUrlWasChanged) {
-      setIsChainIdSet(false)
-    }
-  }, [chainIdFromUrlWasChanged])
+    // Triggering only on changes from URL
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [tradeStateFromUrl])
 
   /**
-   * STEP 2
-   * Set chainId from URL into wallet provider once on page load
-   * It's needed because useWeb3React() returns mainnet chainId by default
+   * On:
+   *  - chainId in URL changes
+   *  - connector changes
+   *  - SPECIAL CASE (rememberedUrlState)
+   *
+   * Note: useEagerlyConnect() changes connectors several times at the beginning
+   *
+   * 1. When chainId in URL is changed, then set it to the provider
+   * 2. If provider's chainId is the same with chainId in URL, then do nothing
+   * 3. When the URL state is remembered, then set it's chainId to the provider
    */
   useLayoutEffect(() => {
-    if (isChainIdSet || !chainIdFromUrl || !currentChainId) return
-    let isSubscribed = true
-
-    setIsChainIdSet(true)
+    if (!providerChainId || providerChainId === currentChainId) return
 
-    switchChain(connector, chainIdFromUrl)
-      .catch((error: Error) => {
-        // We are ignoring Gnosis safe context error
-        // Because it's a normal situation when we are not in Gnosis safe App
-        if (error.name === 'NoSafeContext') return
+    const targetChainId = rememberedUrlState?.chainId || currentChainId
 
-        console.error('Network switching error: ', error)
-      })
-      .finally(() => {
-        if (!isSubscribed) return
+    switchChain(connector, targetChainId).catch((error: Error) => {
+      // We are ignoring Gnosis safe context error
+      // Because it's a normal situation when we are not in Gnosis safe App
+      if (error.name === 'NoSafeContext') return
 
-        updateStateAndNavigate()
-      })
+      console.error('Network switching error: ', error)
+    })
 
-    return () => {
-      isSubscribed = false
-    }
-  }, [isChainIdSet, chainIdFromUrl, connector, currentChainId, updateStateAndNavigate])
+    console.debug('[TRADE STATE]', 'Set chainId to provider', { connector, urlChainId })
+    // Triggering only when chainId in URL is changes and when connector is changed
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [connector, urlChainId, rememberedUrlState])
```

Comment:
Not depending on `providerChainId`?

Reply:
Yes, it depends on only connector, urlChainId, rememberedUrlState